mcbackend/adapters/hopsy.py

```diff
@@ -0,0 +1,76 @@
+import numpy
+import numpy as np
+
+import hopsy
+from typing import Dict, List, Optional, Union
+import hagelkorn
+
+from mcbackend.meta import Coordinate, DataVariable, Variable
+
+from ..core import Backend, Chain, Run, RunMeta
+
+
+class TraceBackend(hopsy.Backend):
+    """Adapter to create a hospy backend from any McBackend."""
+
+    supports_sampler_stats = True
+
+    def __init__(self, backend: Backend, name: str = None):
+        super().__init__(name)
+        self.run_id = hagelkorn.random(digits=6)
+        print(f"Backend run id: {self.run_id}")
+        self._backend: Backend = backend
+
+        # Names
+        self.var_names = None
+
+        # Sessions created from the underlying backend
+        self._run: Optional[Run] = None
+        self._chains: Optional[List[Chain]] = None
+
+    def setup(
+            self,
+            n_chains: int,
+            n_samples: int,
+            n_dim: int,
+            meta_names: List[str],
+            meta_shapes: List[List[int]],
```

Comment:
```suggestion
            meta_names: Sequence[str],
            meta_shapes: Sequence[Sequence[int]],
```
Type annotations of parameters should be covariant (generic) if possible (see [mypy documentation](https://mypy.readthedocs.io/en/stable/common_issues.html#invariance-vs-covariance).

The rule of thumb is: Generic types for inputs, exact types for outputs/attributes.

You have the same issue in https://jugit.fz-juelich.de/IBG-1/ModSim/hopsy/-/blob/develop/src/hopsy/misc.py#L341